warehouse/query-core/src/main/java/datawave/query/jexl/JexlNodeBuilder.java

```diff
@@ -0,0 +1,637 @@
+import org.apache.commons.jexl2.parser.ASTIdentifier;
+import org.apache.commons.jexl2.parser.ASTIfStatement;
+import org.apache.commons.jexl2.parser.ASTJexlScript;
+import org.apache.commons.jexl2.parser.ASTLENode;
+import org.apache.commons.jexl2.parser.ASTLTNode;
+import org.apache.commons.jexl2.parser.ASTMapEntry;
+import org.apache.commons.jexl2.parser.ASTMethodNode;
+import org.apache.commons.jexl2.parser.ASTModNode;
+import org.apache.commons.jexl2.parser.ASTMulNode;
+import org.apache.commons.jexl2.parser.ASTNENode;
+import org.apache.commons.jexl2.parser.ASTNRNode;
+import org.apache.commons.jexl2.parser.ASTNotNode;
+import org.apache.commons.jexl2.parser.ASTNullLiteral;
+import org.apache.commons.jexl2.parser.ASTNumberLiteral;
+import org.apache.commons.jexl2.parser.ASTOrNode;
+import org.apache.commons.jexl2.parser.ASTReference;
+import org.apache.commons.jexl2.parser.ASTReferenceExpression;
+import org.apache.commons.jexl2.parser.ASTReturnStatement;
+import org.apache.commons.jexl2.parser.ASTSizeFunction;
+import org.apache.commons.jexl2.parser.ASTSizeMethod;
+import org.apache.commons.jexl2.parser.ASTStringLiteral;
+import org.apache.commons.jexl2.parser.ASTTernaryNode;
+import org.apache.commons.jexl2.parser.ASTTrueNode;
+import org.apache.commons.jexl2.parser.ASTUnaryMinusNode;
+import org.apache.commons.jexl2.parser.ASTVar;
+import org.apache.commons.jexl2.parser.ASTWhileStatement;
+import org.apache.commons.jexl2.parser.JexlNode;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.function.Supplier;
+
+/**
+ * A builder class that supports fluent-style building of {@link JexlNode} instances.
+ * 
+ * @param <T>
+ *            the {@link JexlNode} type
+ */
+public class JexlNodeBuilder<T extends JexlNode> {
+    
+    private final Supplier<T> constructor;
+    private final List<Supplier<? extends JexlNode>> children = new ArrayList<>();
+    private String image;
+    private Object value;
+    
+    public JexlNodeBuilder(JexlNodeInstance<T> instance) {
+        this.constructor = instance.getConstructor();
+    }
+    
+    /**
+     * Set the image for the node this builder will create.
+     * 
+     * @param image
+     *            the image
+     * @return this builder
+     */
+    public JexlNodeBuilder<T> withImage(String image) {
+        this.image = image;
+        return this;
+    }
+    
+    /**
+     * Set the value for the node this builder will create.
+     * 
+     * @param value
+     *            the value
+     * @return this builder
+     */
+    public JexlNodeBuilder<T> withValue(Object value) {
+        this.value = value;
+        return this;
+    }
+    
+    /**
+     * Add a child node to this builder.
+     * 
+     * @param node
+     *            the child node
+     * @return this builder
+     */
+    public JexlNodeBuilder<T> withChild(JexlNode node) {
+        children.add(() -> node);
+        return this;
+    }
```

Comment:
I would consider removing this.  IMO whoever uses the builder should have their children sorted out before they invoke the builder.

Reply:
Children are stored as an array in a JexlNode, yes, but not in a JexlNodeBuilder. They're added to the list variable `JexlNodeBuilder.children`, and a finalized array of children is created when `JexlNodeBuilder.build()` is called. Your example
```
builder
  .withChildren(new JexlNode[] {
    child1,
    child2,
    child3})
  .build()
```
would result in each child in the array being added to the underlying list, not in it being directly saved in the resulting JexlNode. 

Although only supporting passing in an array of children would be very efficient, I would argue the goal of utilizing the builder pattern here is not just efficiency, but also convenience, and that there's a balance to be found. I prefer to retain the `withChild(JexlNode child)` method for now, and if it turns out not to be used in practice, remove it in a future optimization update.